src/typechecker/environment.lisp

```diff
@@ -15,7 +15,11 @@
                 (immutable-map-data env))))
 
 (defmethod type-variables ((env value-environment))
-  (remove-duplicates (mapcan #'type-variables (fset:convert 'list (fset:range (immutable-map-data env)))) :test #'equalp))
+  (let ((out nil))
+    (fset:do-map (name type (immutable-map-data env))
+      (declare (ignore name))
+      (setf out (append (type-variables type) out)))
+    (remove-duplicates out :test #'equalp)))
```

Comment:
Any chance at a more specific equality predicate?

Reply:
We don't have one defined, and equalp is used everywhere